src/utils/theme/theme.provider.tsx

```diff
@@ -0,0 +1,249 @@
+  lastThemeStateRef.current = themeState;
+
+  const setTheme = useCallback(async (themeMode: ThemeMode) => {
+    // TODO: We might want to add a value "use dApp preference" to clear the setting.
+
+    if (!themeMode && !THEME_MODES.includes(themeMode)) throw new Error(`Invalid theme "${themeMode}"`);
+
+    if (themeMode === lastThemeStateRef.current.themeMode) return;
+
+    window.clearTimeout(setThemeSettingAndStateTimeoutId.current);
+
+    async function setThemeSettingAndState() {
+      try {
+        const nextThemeState = {
+          themeMode,
+          displayTheme: resolveThemeMode(themeMode),
+        };
+
+        // Update state immediately to avoid waiting for the useEffect above to run:
+        setThemeState(nextThemeState);
+
+        // Persist preference:
+        await setThemeSetting(themeMode);
+
+        return nextThemeState;
+      } catch (err) {
+        console.error(`Error changing theme to "${themeMode}":`, err);
+
+        // Revert change:
+        setThemeState(lastThemeStateRef.current);
+      }
+
+      return null;
+    }
+
+    const cover = document.getElementById("cover");
+    const displayTheme = resolveThemeMode(themeMode);
+
+    if (cover && displayTheme !== lastThemeStateRef.current.displayTheme) {
+      // If the resolved theme changes and we have a cover element, we want to prevent users from seeing a mix of
+      // properties transitioning to their new value and some other ones abruptly changing, so:
+
+      return new Promise<ThemeContextState>((resolve) => {
+        // 1. We show the cover:
+        cover.removeAttribute("aria-hidden");
+
+        // 2. Wait for its fade-in transition to finish:
+        setThemeSettingAndStateTimeoutId.current = window.setTimeout(() => {
+          // 3. Change the theme:
+          const updatedThemeState = setThemeSettingAndState();
+
+          // 4. Wait for potential transition on properties affected by the theme change to finish:
+          setThemeSettingAndStateTimeoutId.current = window.setTimeout(() => {
+            // Hide it again:
+            cover.setAttribute("aria-hidden", "true");
+
+            resolve(updatedThemeState);
+          }, 230);
+        }, 230);
+      });
+    }
+
+    // Otherwise, we update the preference and state straight away:
+    return setThemeSettingAndState();
+  }, []);
+
+  // Listen for dark mode browser preference changes if the current themeMode is "system":
+
+  useEffect(() => {
+    if (themeMode !== "system") return;
+
+    function handleDarkModePreferenceChange(e: MediaQueryListEvent) {
+      setThemeState(getThemeState());
+    }
+
+    darkModePreference.addEventListener("change", handleDarkModePreferenceChange);
+
+    return () => darkModePreference.removeEventListener("change", handleDarkModePreferenceChange);
+  }, [themeMode, getThemeState]);
+
+  // (ONLY CONNECT, NOT BE) Add `data-theme` attribute based on resolved theme (light or dark, not system):
```

Comment:
Comment doesn't match the useEffect below.

Reply:
What do you mean? `import.meta.env?.VITE_IS_EMBEDDED_APP` is set to `"1"` only for Connect, so in BE both `useEffect`s will simply return here: `if (import.meta.env?.VITE_IS_EMBEDDED_APP !== "1") return;`